iriscasttools/iriscasttools/utils.py

```diff
@@ -0,0 +1,224 @@
+"""
+Provides utility functions to collect energy usage information
+"""
+import subprocess
+import time
+import os
+from pathlib import Path
+import re
+from typing import Dict
+import logging
+import datetime
+
+logger = logging.getLogger(__name__)
+
+
+def retry(
+    retry_on: (Exception,) = None,
+    retries: int = 3,
+    delay: int = 3,
+    backoff: int = 2,
+    retry_logger: logging.Logger = None,
+):
+    """A retry decorator
+
+    Keyword Arguments:
+        retry_on: a set of exceptions that when any occur, will trigger a retry
+        retries: number of retries to perform before returning None
+        delay: number of seconds delay before next retry
+        backoff: a factor to increase delay after every fail
+        retry_logger: logger to log failed attempts
+    """
+
+    def decorator(func):
+        def inner(*args, **kwargs):
+            res = None
+            for i in range(retries + 1):
+                try:
+                    res = func(*args, **kwargs)
```

Comment:
Can we add debug logging for what args and kwargs we're about to try before
and something to debug log the return code